activesupport/lib/active_support/duration.rb

```diff
@@ -89,21 +89,83 @@ def ago(time = ::Time.current)
         to_sentence(locale: ::I18n.default_locale)
     end
 
+    # JSON representation should be a string formatted as ISO 8601 Duration, as recommended by Google JSON Style Guide:
+    # https://google-styleguide.googlecode.com/svn/trunk/jsoncstyleguide.xml?showone=Time_Duration_Property_Values#Time_Duration_Property_Values
     def as_json(options = nil) #:nodoc:
-      to_i
+      iso8601(precision: options && options[:precision])
     end
 
     def respond_to_missing?(method, include_private=false) #:nodoc:
       @value.respond_to?(method, include_private)
     end
 
+    # Creates a new Duration from string formatted according to ISO 8601 Duration.
+    #
+    # See http://en.wikipedia.org/wiki/ISO_8601#Durations
+    # This method allows negative parts to be present in pattern.
+    # If invalid string is provided, it will raise +ActiveSupport::Duration::ISO8601DurationParser::ParsingError+.
+    def self.parse!(iso8601duration)
+      require 'active_support/duration/iso8601_parser' unless defined?(ISO8601Parser)
+      parts = ISO8601Parser.new(iso8601duration).parts
+      time  = ::Time.now
+      new(time.advance(parts) - time, parts)
+    end
+
+    # Creates a new Duration from string formatted according to ISO 8601 Duration.
+    #
+    # See http://en.wikipedia.org/wiki/ISO_8601#Durations
+    # This method allows negative parts to be present in pattern.
+    # If invalid string is provided, nil will be returned.
+    def self.parse(iso8601duration)
+      parse!(iso8601duration)
+    rescue ISO8601Parser::ParsingError
+      nil
+    end
+
+    # Build ISO 8601 Duration string for this duration.
+    # The +precision+ parameter can be used to limit seconds' precision of duration.
+    def iso8601(precision: nil)
+      output, sign = 'P', ''
+      parts = normalized_parts
+      # If all parts are negative - let's output negative duration
+      if parts.values.compact.all?{|v| v < 0 }
+        sign = '-'
+        parts = parts.inject({}) {|p,(k,v)| p[k] = -v; p }
+      end
+      # Building output string
+      output << "#{parts[:years]}Y"   if parts[:years]
+      output << "#{parts[:months]}M"  if parts[:months]
+      output << "#{parts[:weeks]}W"   if parts[:weeks]
+      output << "#{parts[:days]}D"    if parts[:days]
+      time = ''
+      time << "#{parts[:hours]}H"     if parts[:hours]
+      time << "#{parts[:minutes]}M"   if parts[:minutes]
+      if parts[:seconds]
+        time << "#{sprintf(precision ? "%0.0#{precision}f" : '%g', parts[:seconds])}S"
+      end
+      output << "T#{time}"  if time.present?
+      "#{sign}#{output}"
+    end
+
     delegate :<=>, to: :value
 
+    # Return duration's parts summarized (as they can become repetitive due to addition, etc)
+    # Also removes zero parts as not significant
+    def normalized_parts
+      self.parts.inject({}) do |p,(k,v)|
+        unless v.zero?
+          p[k] ||= 0
+          p[k]  += v
+        end
+        p
+      end
```

Comment:
I agree on the `each_with_object`, but I did advise against the default earlier because it was set to nil after the iteration (and therefore felt like a hack to me).